### PR TITLE
fix(gc): drop stale shape entries at recycled keys-array addresses

### PR DESCRIPTION
**Pull request by jdalton**
## Summary

Prevent weak shape indices and descriptors from surviving when a dead keys-array address has been recycled by a different GC object type. This is scoped shape-table hardening found during the seeded-GC investigation; the observed `node-machine-id` / `Function("return this")()` TypeError had a separate unrooted-global cause fixed by #8333.

## Changes

- Validate tracked shape-key objects during dead-owner pruning and reject non-array/lazy-array tenants even when the replacement object is live.
- Validate descriptor keys after metadata visitation, and drop descriptors or indices rather than rekeying them through a replacement object's forwarding record.
- Remove the unused temporary `diagnose_shape_rekey_gap` diagnostic.
- Add exact Eden-address reuse coverage with a real copying minor, nonzero-copy and movement gates, a sabotage arm, and focused tests for both review findings.
- Add `changelog.d/8324-stale-shape-recycled-address.md`.

## Related issue

n/a — follow-up hardening discovered while diagnosing the failure resolved by #8333.

## Test plan

- `cargo test -p perry-runtime shape_keys_descriptor_edge -- --nocapture` (8 passed)
- `cargo test -p perry-runtime shapes -- --nocapture` (24 passed)
- `cargo test -p perry-runtime test_shape_ -- --nocapture` (4 passed)
- `cargo check -p perry-runtime`
- `cargo fmt --all -- --check`
- `python3 scripts/shape_descriptor_census.py`
- `python3 scripts/addr_class_inventory.py`
- `python3 scripts/gc_rekeyed_key_tables.py`
- `./scripts/check_file_size.sh`
- `BASE_SHA=origin/main SKIP_COMPILE_GATES=1 ./scripts/run_lint_gates.sh` (47 passing; the one initial census failure was fixed and its gate rerun successfully)

- [ ] `cargo build --release` clean
- [ ] Full affected-workspace test scope passes locally
- [x] Added a `#[test]` regression in the affected runtime crate
- [x] Runtime API docs unchanged; no user-facing API update needed
- [x] No platform UI backend touched

## Screenshots / output

n/a

## Checklist

- [x] I have NOT bumped the workspace version or edited CLAUDE.md / CHANGELOG.md
- [x] My commit follows the repository's `fix:` prefix convention
- [x] I've read `CONTRIBUTING.md` and agree to the Code of Conduct

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved runtime handling of moved objects to prevent stale shape and metadata references.
  - Removed invalid entries when memory addresses are recycled for different object types.
  - Rebuilt internal indexes after object forwarding to keep shape information consistent.
  - Improved garbage collection reliability when memory is reused across different object types.

### CODE REVIEW

**Comment by proggeramlug** [comment]
Good investigation — the instrumentation evidence (2 stale `indices`, 4 stale
`descriptor` keys, at addresses moved as `obj_type=3`/`4` rather than
`GC_TYPE_ARRAY`, with every other side table clean) is exactly the right way to
localise this class. Holding it for three reasons, one of which is timing rather
than anything wrong with the code.

## 1. It collides with #8313, and one of your three fixes becomes moot

#8313 (`perf(object): shrink common objects to 40 bytes`, closes #8047) deletes
the `ObjectHeader::keys_array` mirror entirely. I checked both branches:

| your fix | target | after #8313 |
|---|---|---|
| 3. `gc_keys_array_slot` mirror-refresh guard | `fn gc_keys_array_slot` | **gone** — 0 occurrences; the function does not exist |
| 1. `prune_dead_shape_keys` type check | `fn prune_dead_shape_keys` | survives |
| 2. `scan_shape_table_rekey_mut` recycled-entry drop | `fn scan_shape_table_rekey_mut` | survives |

`git merge-tree` reports a real conflict between the two branches. #8313 is
verified and about to land, so the ask is: rebase on top of it, drop fix 3 (the
mirror it guards no longer exists — which also removes the corruption path your
summary describes, "refreshes the object's `keys_array` mirror from the stale
descriptor"), and keep 1 and 2, which are still needed because the shape tables
are still keyed by keys-array address.

## 2. No regression test

This is the category CLAUDE.md is most emphatic about: a stale-entry bug is
invisible at collection time and surfaces cycles later somewhere unrelated, so
"the suite is green" says nothing. #8313's `gc::tests::shape_keys_descriptor_edge`
is the shape to copy — it runs a real copying minor, gates on
`copied_objects > 0` **and** on the receiver's address actually changing, asserts
a discriminating quantity, and includes a sabotage arm that re-runs the identical
workload with the edge suppressed and asserts the detector notices. Without that
last arm a green run cannot distinguish "the fix works" from "nothing was
exercised".

Your three seeds (8/11/22) already sound like the makings of a fixture.

## 3. `diagnose_shape_rekey_gap` is dead code

It is `#[allow(dead_code)]`, `eprintln!`-based, labelled "TEMPORARY DIAGNOSTIC",
and has no caller anywhere in the diff. Please drop it before merge — the repo's
standing rule is that an unexercised path is a configuration nobody has verified,
and a diagnostic that no test invokes rots immediately.

## On the honest note in your summary

> the TypeError still fires on seeds 8/11/22 despite all scanned side tables
> showing 0 stale entries

Thank you for stating that plainly — it is the most useful line in the PR. It
does mean this narrows a staleness class rather than fixing the reported bug, so
please scope the title/changelog accordingly (a `changelog.d/8324-*.md` fragment
is also still missing). Worth re-checking whether the remaining symptom survives
*on top of* #8313, since removing the mirror deletes one of the two ways a stale
descriptor could reach a lookup.

**Comment by proggeramlug** [comment]
Update that may resolve your open question. Your summary ended with:

> the TypeError still fires on seeds 8/11/22 despite all scanned side tables
> showing 0 stale entries. The `undefined` may come from an additional path not
> covered by the current instrumentation (possibly the `Function("return this")()`
> interpreter path in node-machine-id).

That hunch looks right. #8333 (now merged) found and fixed exactly that path:
`dyn_function_from_strings` read the global via `js_get_global_this()` but did
not root it before `env_new_root()`, which allocates. A copying minor there
evacuates the global singleton — `THREAD_GLOBAL_THIS` is a registered root and
gets rewritten, but the raw local does not, so the stale pointer flows into
capture slots 3 and 4. On `Function("return this")()` the sloppy-mode
`this = global` then reads back undefined, and
`Object.getPrototypeOf(undefined)` throws
`TypeError: Cannot convert undefined or null to object` — the symptom you were
chasing.

So the remaining source was an unrooted register, not a stale table. Worth
re-running your seeds on current `main` before doing more work here: it is
plausible that the shape-table staleness you found was real but not what was
producing the TypeError.

Two things from my earlier review still stand if you take this further:

- #8313 has landed, so `gc_keys_array_slot` no longer exists — that fix of yours
  is moot now, while `prune_dead_shape_keys` and `scan_shape_table_rekey_mut`
  still apply.
- The `#[allow(dead_code)]` `diagnose_shape_rekey_gap` should come out, and a
  fix in this area wants a fixture with a sabotage arm (see
  `gc::tests::shape_keys_descriptor_edge`, which #8313 added).

If the seeds come back clean on current `main`, this may be worth closing rather
than rebasing.

**Comment by proggeramlug** [comment]
Addressed the review feedback in `7f84d81d6`:

- rebased onto current `main` (including #8313 and #8333); the removed `gc_keys_array_slot` mirror path is not part of this PR;
- removed the unused `diagnose_shape_rekey_gap` diagnostic;
- validate descriptor keys after metadata visitation and remove any readable non-array/lazy-array result;
- removed the `GC_FLAG_FORWARDED` requirement for stale index cleanup;
- added an exact Eden-address reuse regression using a real copying minor, with `copied_objects > 0`, replacement-address movement, and a sabotage arm, plus focused tests for the non-forwarded prune and post-visit rewrite cases;
- added the `changelog.d/8324-stale-shape-recycled-address.md` fragment;
- rewrote the PR description to scope this as shape-table hardening and credit #8333 for the separate `Function("return this")()` / `node-machine-id` TypeError root cause.

Local validation is listed in the updated PR description. The focused descriptor-edge suite passes 8/8, the broader shape suite passes 24/24, the shape GC migration subset passes 4/4, and the relevant lint/census gates pass.

File: crates/perry-runtime/src/gc/tests/shape_keys_descriptor_edge.rs [hunk 1]
```diff
@@ -269,3 +269,172 @@ fn a_keys_array_whose_last_carrier_died_is_still_reclaimed() {
...
+        !shapes::test_shape_entry_exists(from) && !shapes::test_shape_entry_exists(to),
+        "the index must be dropped instead of following a closure's forwarding record"
+    );
+    assert!(
+        shapes::test_shape_id_for_keys(from).is_none()
+            && shapes::test_shape_id_for_keys(to).is_none(),
+        "the descriptor must validate the rewritten address before publishing it"
+    );
+    shapes::test_clear_shape_table();
+}
```
[the remark this answers is not in the document]
Comment (jdalton) [reply crates/perry-runtime/src/gc/tests/shape_keys_descriptor_edge.rs:440]: The recycled-address edge case this flags is exactly what `shape_keys_descriptor_edge.rs` exercises — the test file exists to catch a stale address rewritten through a forwarded object. The shape-descriptor path (shapes.rs) re-validates the key bytes on every publish, and the descriptor is the source of truth, not `ObjectHeader::keys_array` (which these tests suppress). The invariant is documented in the module header (shapes.rs:3-14) + the test header (shape_keys_descriptor_edge.rs:6-23).
